### ISSUE TITLE
Possible memory leak when using MetadataPropertyHandling.ReadAhead?

### ISSUE THREAD

**Issue by enrique-avalon** [JamesNK/Newtonsoft.Json#871]
Hi,

I'm deserializing json-data for a rather large graph (the raw json is 42MB). When I'm setting MetadataPropertyHandling to read-ahead about 700MB of memory is consumed and the deserialization take about 35 seconds. Whatever is taking up the memory is holding on to it until the application is terminated. In this application I really didn't need the read-ahead setting and by turning it off the memory consumption for deserialization went down to almost nothing (in comparison) and the entire deserialization process took about 4 seconds. I can understand why it takes way longer to read ahead - but the memory that never gets released is what I'm curious about.

Using version 8.0.3

**Comment by JamesNK** [comment]
I think I have an idea what is causing it.

Does your deserialized object have any properties with type of object or JToken on it?

**Comment by JamesNK** [comment]
My idea didn't work.

Are you holding onto what you deserialized?
Is there a way I can repo your issue?

**Comment by enrique-avalon** [comment]
Yes I have the data, not at my computer right now but will provide data and code to reproduce the error as soon as possible (a couple of hours or so).

**Comment by enrique-avalon** [comment]
Hi again,

Sorry for the late response. I had to anonymize the data. Anyhow, I had to reference parts of our domain model and I can't really attach that here, I have a project ready where I have reproduced the issue but I'd rather e-mail it to you, is that possible?

**Comment by JamesNK** [comment]
Sure. james at newtonking.com

**Comment by JamesNK** [comment]
I'm afraid I couldn't replicate it. I can see a lot of memory usage while the deserialization is happening but it is all released once complete. There is an increase of 13 megabytes from the deserialized objects.

I also tested against the latest source code and there were some significant memory allocation improvements from optimizations I had already made:
8.0.3 used 6.5 gigabytes
Latest used 0.5 gigabytes
